File: BinaryTree/csharp/AllTreePaths.cs
using System.Collections.Generic;

namespace BinaryTreeSolutions;

public static class AllTreePaths
{
    public static IList<IList<int>> Solve(TreeNode? root)
    {
        var paths = new List<IList<int>>();
        Traverse(root, new List<int>(), paths);
        return paths;
    }

    private static void Traverse(TreeNode? node, IList<int> current, IList<IList<int>> paths)
    {
        if (node is null)
        {
            return;
        }

        current.Add(node.Val);
        if (node.Left is null && node.Right is null)
        {
            paths.Add(new List<int>(current));
        }
        else
        {
            Traverse(node.Left, current, paths);
            Traverse(node.Right, current, paths);
        }
        current.RemoveAt(current.Count - 1);
    }
}

File: BinaryTree/java/allTreePaths.java
import java.util.ArrayList;
import java.util.List;

class AllTreePaths {
  public static List<List<Integer>> allTreePaths(TreeNode root) {
    List<List<Integer>> paths = new ArrayList<>();
    if (root == null) {
      return paths;
    }
    traverse(root, new ArrayList<>(), paths);
    return paths;
  }

  private static void traverse(TreeNode node, List<Integer> current, List<List<Integer>> paths) {
    if (node == null) {
      return;
    }

    current.add(node.val);
    if (node.left == null && node.right == null) {
      paths.add(new ArrayList<>(current));
    } else {
      traverse(node.left, current, paths);
      traverse(node.right, current, paths);
    }
    current.remove(current.size() - 1);
  }

  public static void run() {
    // intentionally empty
  }
}
